astropy/io/ascii/cparser.pyx

```diff
@@ -343,7 +341,12 @@ cdef class CParser:
                 raise core.InconsistentTableError('No data lines found, C reader '
                                             'cannot autogenerate column names')
             # auto-generate names
-            self.names = ['col{0}'.format(i + 1) for i in range(self.width)]
+            self.header_names = ['col{0}'.format(i + 1) for i in range(self.width)]
+
+        if self.names:
+            self.width = len(self.names)
+        else:
+            self.names = self.header_names
```

Comment:
I don't know much about the fast readers. Is there a reason that this does not read:
```
if not self.names:
    self.names = self.header_names
self.width = len(self.names)
```